File: src/test/java/ru/anseranser/peshki/engine/BoardTest.java
package ru.anseranser.peshki.engine;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class BoardTest {

    private GameConfig config;
    private Board board;

    @BeforeEach
    void setUp() {
        config = GameConfig.DEFAULT;
        board = new Board(config);
    }

    @Test
    void createsCorrectNumberOfPlayers() {
        assertEquals(4, board.getPlayers().size());
    }

    @Test
    void eachPlayerHasCorner() {
        for (Player player : board.getPlayers()) {
            assertNotNull(board.getCorner(player));
            assertEquals(Cell.CellType.CORNER, board.getCorner(player).getCellType());
        }
    }

    @Test
    void cornersBelongToCorrectPlayers() {
        for (Player player : board.getPlayers()) {
            assertEquals(player, board.getCorner(player).getOwner());
        }
    }

    @Test
    void fieldCellsFormRing() {
        Player p1 = board.getPlayers().get(0);
        Cell corner = board.getCorner(p1);
        Cell current = corner;

        int count = 0;
        do {
            current = current.getNextFieldCell();
            count++;
        } while (current != corner);

        // Ring includes 4 corners + fieldLength field cells = 28 total nodes
        assertEquals(config.numberOfPlayers() + config.fieldLength(), count);
    }

    @Test
    void homeCellsHaveCorrectCount() {
        for (Player player : board.getPlayers()) {
            Cell corner = board.getCorner(player);
            int homeCount = 0;
            Cell current = corner.getNextHomeCell();
            while (current != null) {
                homeCount++;
                assertEquals(Cell.CellType.HOME, current.getCellType());
                assertEquals(player, current.getOwner());
                current = current.getNextHomeCell();
            }
            assertEquals(config.homeLength(), homeCount);
        }
    }

    @Test
    void homeCellsBranchFromCorners() {
        for (Player player : board.getPlayers()) {
            Cell corner = board.getCorner(player);
            assertNotNull(corner.getNextHomeCell());
            assertEquals(Cell.CellType.HOME, corner.getNextHomeCell().getCellType());
        }
    }

    @Test
    void allCellsAreUnique() {
        var allCells = board.getAllCells();
        assertEquals(allCells.size(), allCells.stream().distinct().count());
    }
}

File: web/tests/Board.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { DEFAULT_CONFIG } from '../src/engine/GameConfig';
import { Board } from '../src/engine/Board';

describe('Board', () => {
  let board: Board;

  beforeEach(() => {
    board = new Board(DEFAULT_CONFIG);
  });

  it('creates correct number of players', () => {
    expect(board.players.length).toBe(4);
  });

  it('each player has corner', () => {
    for (const player of board.players) {
      expect(board.getCorner(player)).toBeDefined();
    }
  });

  it('corners belong to correct players', () => {
    for (const player of board.players) {
      expect(board.getCorner(player).owner).toBe(player);
    }
  });

  it('field cells form ring', () => {
    const corner = board.getCorner(board.players[0]);
    let current = corner;
    let count = 0;
    do {
      current = current.nextFieldCell!;
      count++;
    } while (current !== corner);
    expect(count).toBe(28);
  });

  it('each player has home cells', () => {
    for (const player of board.players) {
      const corner = board.getCorner(player);
      let home = corner.nextHomeCell;
      let count = 0;
      while (home !== null) {
        count++;
        home = home.nextHomeCell;
      }
      expect(count).toBe(3);
    }
  });

  it('all cells are unique', () => {
    const cells = board.getAllCells();
    const unique = new Set(cells);
    expect(unique.size).toBe(cells.length);
  });
});
